File: stripe-dashboard-refund-not-synced/python/sync_refunds.py
"""Record Stripe dashboard refunds that never synced to WooCommerce.
Uses api_refund false, so it never refunds the customer twice.
Run on a schedule. Safe to run again and again.

Guide: https://www.allanninal.dev/woocommerce/stripe-dashboard-refund-not-synced/
"""
import os
import time
import logging
import stripe
import requests
from requests.auth import HTTPBasicAuth

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("sync_refunds")

stripe.api_key = os.environ["STRIPE_SECRET_KEY"]
WOO_URL = os.environ["WOO_STORE_URL"].rstrip("/")
AUTH = HTTPBasicAuth(os.environ["WOO_CONSUMER_KEY"], os.environ["WOO_CONSUMER_SECRET"])
LOOKBACK_HOURS = int(os.environ.get("LOOKBACK_HOURS", "72"))
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"


def recent_refunded_charges(lookback_hours):
    since = int(time.time()) - lookback_hours * 3600
    seen = set()
    for refund in stripe.Refund.list(limit=100, created={"gte": since}).auto_paging_iter():
        charge_id = refund.get("charge")
        if not charge_id or charge_id in seen:
            continue
        seen.add(charge_id)
        charge = stripe.Charge.retrieve(charge_id)
        order_id = (charge.get("metadata") or {}).get("order_id")
        if order_id:
            yield order_id, charge


def wc_refunds(order_id):
    r = requests.get(f"{WOO_URL}/wp-json/wc/v3/orders/{order_id}/refunds", auth=AUTH, timeout=30)
    if r.status_code == 404:
        return None
    r.raise_for_status()
    return r.json()


def wc_refunded_minor(refunds):
    return sum(round(abs(float(r["amount"])) * 100) for r in refunds)


def missing_refund_minor(stripe_refunded_minor, refunds):
    gap = stripe_refunded_minor - wc_refunded_minor(refunds)
    return gap if gap > 1 else 0


def record_refund(order_id, amount_minor):
    requests.post(
        f"{WOO_URL}/wp-json/wc/v3/orders/{order_id}/refunds",
        json={
            "amount": f"{amount_minor / 100:.2f}",
            "reason": "Recorded from a Stripe dashboard refund. No money moved on Stripe.",
            "api_refund": False,
        },
        auth=AUTH, timeout=30,
    ).raise_for_status()


def run():
    recorded = 0
    for order_id, charge in recent_refunded_charges(LOOKBACK_HOURS):
        refunds = wc_refunds(order_id)
        if refunds is None:
            log.warning("Charge for order %s but the order is missing in Woo", order_id)
            continue
        missing = missing_refund_minor(charge["amount_refunded"], refunds)
        if not missing:
            continue
        log.info("Order %s: Stripe refunded %s more than Woo has. %s",
                 order_id, missing, "would record" if DRY_RUN else "recording")
        if not DRY_RUN:
            record_refund(order_id, missing)
        recorded += 1
    log.info("Done. %d refund(s) %s.", recorded, "to record" if DRY_RUN else "recorded")


if __name__ == "__main__":
    run()

File: stripe-dashboard-refund-not-synced/node/sync-refunds.js
/**
 * Record Stripe dashboard refunds that never synced to WooCommerce.
 * Uses api_refund false, so it never refunds the customer twice.
 * Run on a schedule. Safe to run again and again.
 *
 * Guide: https://www.allanninal.dev/woocommerce/stripe-dashboard-refund-not-synced/
 */
import Stripe from "stripe";
import { pathToFileURL } from "node:url";

const stripe = new Stripe(process.env.STRIPE_SECRET_KEY || "sk_test_dummy");
const WOO_URL = (process.env.WOO_STORE_URL || "https://example.com").replace(/\/$/, "");
const AUTH = "Basic " + Buffer.from(
  `${process.env.WOO_CONSUMER_KEY || "ck_dummy"}:${process.env.WOO_CONSUMER_SECRET || "cs_dummy"}`
).toString("base64");
const LOOKBACK_HOURS = Number(process.env.LOOKBACK_HOURS || 72);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

async function woo(path, options = {}) {
  const res = await fetch(`${WOO_URL}/wp-json/wc/v3${path}`, {
    ...options,
    headers: { "Content-Type": "application/json", Authorization: AUTH, ...(options.headers || {}) },
  });
  if (res.status === 404) return null;
  if (!res.ok) throw new Error(`Woo ${path} returned ${res.status}`);
  return res.json();
}

async function* recentRefundedCharges(lookbackHours) {
  const since = Math.floor(Date.now() / 1000) - lookbackHours * 3600;
  const seen = new Set();
  for await (const refund of stripe.refunds.list({ limit: 100, created: { gte: since } })) {
    const chargeId = refund.charge;
    if (!chargeId || seen.has(chargeId)) continue;
    seen.add(chargeId);
    const charge = await stripe.charges.retrieve(chargeId);
    const orderId = (charge.metadata || {}).order_id;
    if (orderId) yield { orderId, charge };
  }
}

export function wcRefundedMinor(refunds) {
  return refunds.reduce((sum, r) => sum + Math.round(Math.abs(parseFloat(r.amount)) * 100), 0);
}

export function missingRefundMinor(stripeRefundedMinor, refunds) {
  const gap = stripeRefundedMinor - wcRefundedMinor(refunds);
  return gap > 1 ? gap : 0;
}

async function recordRefund(orderId, amountMinor) {
  await woo(`/orders/${orderId}/refunds`, {
    method: "POST",
    body: JSON.stringify({
      amount: (amountMinor / 100).toFixed(2),
      reason: "Recorded from a Stripe dashboard refund. No money moved on Stripe.",
      api_refund: false,
    }),
  });
}

export async function run() {
  let recorded = 0;
  for await (const { orderId, charge } of recentRefundedCharges(LOOKBACK_HOURS)) {
    const refunds = await woo(`/orders/${orderId}/refunds`);
    if (refunds === null) { console.warn(`Charge for order ${orderId} but the order is missing in Woo`); continue; }
    const missing = missingRefundMinor(charge.amount_refunded, refunds);
    if (!missing) continue;
    console.log(`Order ${orderId}: Stripe refunded ${missing} more than Woo has. ${DRY_RUN ? "would record" : "recording"}`);
    if (!DRY_RUN) await recordRefund(orderId, missing);
    recorded++;
  }
  console.log(`Done. ${recorded} refund(s) ${DRY_RUN ? "to record" : "recorded"}.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
